### PR TITLE
feat: add configurable checksumming for blob uploads and downloads

### PR DESCRIPTION
**Pull request by andrewsg**

### CODE REVIEW

File: google/cloud/storage/blob.py [hunk 1]
```diff
@@ -1838,10 +1937,16 @@ def _do_resumable_upload(
...
+            checksum=checksum,
         )
 
         while not upload.finished:
-            response = upload.transmit_next_chunk(transport, timeout=timeout)
+            try:
+                response = upload.transmit_next_chunk(transport, timeout=timeout)
+            except resumable_media.DataCorruption:
+                # Attempt to delete the corrupted object.
+                self.delete()
```
Comment (andrewsg) [inline google/cloud/storage/blob.py:1948]: If this fails for any reason, both exceptions will be reported, using the Python exception pattern "During handling of the above exception, another exception occurred"

File: google/cloud/storage/blob.py [hunk 2]
```diff
@@ -2067,6 +2189,19 @@ def upload_from_file(
             Can also be passed as a tuple (connect_timeout, read_timeout).
             See :meth:`requests.Session.request` documentation for details.
 
+        :type checksum: str
+        :param checksum:
+            (Optional) The type of checksum to compute to verify
+            the integrity of the object. If the upload is completed in a single
```
Comment (andrewsg) [inline google/cloud/storage/blob.py:2195]: This "if the upload is completed in a single request" / "if the upload is too large" is undeniably ugly and hopefully we can implement server-side handling of resumable upload checksums to resolve that soon

**Comment by andrewsg** [comment]
Recommend hold off on merging until google-resumable-media releases to 1.0 (and bump up the dependency version number in setup.py here)

File: google/cloud/storage/blob.py (hunk 1 above)
Comment (tseaver) [reply google/cloud/storage/blob.py:1948]: What occurs here if a valid previous generation of the blob exists?

File: tests/system/test_system.py [hunk 3]
```diff
@@ -814,6 +871,33 @@ def test_download_w_generation_match(self):
 
         self.assertEqual(file_contents, stored_contents)
 
+    def test_download_w_failed_crc32c_checksum(self):
+        blob = self.bucket.blob("FailedChecksumBlob")
+        file_contents = b"Hello World"
+        blob.upload_from_string(file_contents)
+        self.case_blobs_to_delete.append(blob)
+
+        temp_filename = tempfile.mktemp()
```
Comment (tseaver) [inline tests/system/test_system.py:880]: I just opened #247 to track removing the unsafe `tempfile.mktemp`.  Here, you would use `tempfile.NamedTemporaryFile`. e.g.:

```python
        with tempfile.NamedTemporaryFile() as temp_f:

            # Intercept the digest processing at the last stage and replace it with garbage
            with mock.patch(
                "google.resumable_media._helpers.prepare_checksum_digest",
                return_value="FFFFFF==",
            ):
                with self.assertRaises(resumable_media.DataCorruption):
                    blob.download_to_filename(temp_f.name, checksum="crc32c")
```

etc.

**Comment by tseaver** [comment]
https://github.com/googleapis/google-resumable-media-python/pull/142 is the PR for bumping GRMP to 1.0.

File: tests/system/test_system.py [hunk 4]
```diff
@@ -555,6 +558,33 @@ def test_large_file_write_from_stream(self):
...
+            md5_hash = md5_hash.encode("utf-8")
+        self.assertEqual(md5_hash, file_data["hash"])
+
+    def test_large_file_write_from_stream_with_failed_checksum(self):
+        blob = self.bucket.blob("LargeFile")
+
+        file_data = self.FILES["big"]
+
+        with open(file_data["path"], "rb") as file_obj:
+            with mock.patch(
```
[the remark this answers is not in the document]
Comment (andrewsg) [reply tests/system/test_system.py:581]: Done

**Comment by andrewsg** [comment]
@tseaver @frankyn PTAL

@tseaver Has the PR you linked made it to release yet? If it's released I'll use this PR to also update setup.py to use the new version.

File: google/cloud/storage/blob.py [hunk 5]
```diff
@@ -1822,6 +1911,16 @@ def _do_resumable_upload(
...
 
+        :type checksum: str
+        :param checksum:
+            (Optional) The type of checksum to compute to verify
+            the integrity of the object. After the upload is complete, the
+            server-computed checksum of the resulting object will be checked
+            and google.resumable_media.common.DataCorruption will be raised on
+            a mismatch. On a failure, the client will attempt to delete the
+            corrupted file from the remote host automatically. Supported values
+            are "md5", "crc32c" and None. The default is None.
```
Comment (frankyn) [inline google/cloud/storage/blob.py:1922]: `On a failure, the client will attempt to delete the corrupted file from the remote host automatically` 

Suggestion:
`On a validation failure, the client will attempt to delete the uploaded object automatically.`

Comment (andrewsg) [reply google/cloud/storage/blob.py:1922]: Done

**Comment by andrewsg** [comment]
Bumped version number. This should be good to merge as soon as final review comes through. Thanks all

**Comment by andrewsg** [comment]
@tseaver PTAL soon if you are available, I've been asked to release. Thanks!

**Comment by tseaver** [comment]
https://pypi.org/project/google-resumable-media/1.0.0/ is released.

**Pull request merged**